common/src/main/kotlin/org/valkyrienskies/mod/common/BlockStateInfoProvider.kt

```diff
@@ -1,42 +1,72 @@
+import net.minecraft.util.registry.RegistryKey
+import net.minecraft.util.registry.SimpleRegistry
 import net.minecraft.world.World
 import org.valkyrienskies.core.game.VSBlockType
-import org.valkyrienskies.core.game.VSBlockType.AIR
-import org.valkyrienskies.core.game.VSBlockType.SOLID
-import org.valkyrienskies.core.game.VSBlockType.WATER
-
-object BlockStateInfoProvider {
-    private fun getBlockStateMass(blockState: BlockState): Double {
-        if (blockState.isAir) return 0.0
-        return 100.0
-    }
+import org.valkyrienskies.mod.common.ValkyrienSkiesMod.MASS_DATAPACK_RESOLVER
+
+// Other mods can then provide weights and types based on their added content
+// NOTE: if we have block's in vs-core we should ask getVSBlock(blockstate: BlockStat): VSBlock since thatd be more handy
+//  altough we might want to allow null properties in VSBlock that is returned since we do want partial data fetching
+// https://github.com/ValkyrienSkies/Valkyrien-Skies-2/issues/25
+interface BlockStateInfoProvider {
+    val priority: Int
+
+    fun getBlockStateMass(blockState: BlockState): Double?
+
+    fun getBlockStateType(blockState: BlockState): VSBlockType?
+}
+
+object BlockStateInfo {
+
+    // registry for mods to add their weights
+    val REGISTRY = SimpleRegistry<BlockStateInfoProvider>(
+        RegistryKey.ofRegistry(Identifier(ValkyrienSkiesMod.MOD_ID, "blockstate_providers")),
+        Lifecycle.experimental()
+    )
 
-    private fun getBlockStateType(blockState: BlockState): VSBlockType {
-        return when {
-            blockState.isAir -> {
-                AIR
-            }
-            blockState.block is FluidBlock -> {
-                WATER
-            }
-            else -> {
-                SOLID
-            }
-        }
+    init {
+        Registry.register(REGISTRY, Identifier(ValkyrienSkiesMod.MOD_ID, "data"), MASS_DATAPACK_RESOLVER)
+        Registry.register(REGISTRY, Identifier(ValkyrienSkiesMod.MOD_ID, "default"), DefaultBlockStateInfoProvider)
     }
 
+    // no clue about these different Int2ObjectMaps just took a random one
+
+    val CACHE = Int2ObjectAVLTreeMap<Pair<Double, VSBlockType>>()
+    // NOTE: this caching can get allot better, ex. default just returns constants so it might be more faster
+    //  if we store that these values do not need to be cached by double and blocktype but just that they use default impl
+
+    // this gets the weight and type provided by providers; or it gets it out of the cache
+
+    private fun get(blockState: BlockState): Pair<Double, VSBlockType> =
+        CACHE.get(Block.getRawIdFromState(blockState)) ?: iterateRegistry(blockState)
+
+    private fun iterateRegistry(blockState: BlockState): Pair<Double, VSBlockType> =
+        Pair(
+            // TODO maybe i should sort the registry; instead of sorting it on the fly :P
+            REGISTRY.sortedByDescending { it.priority }.mapNotNull { it.getBlockStateMass(blockState) }.first(),
```

Comment:
Yeah sorting it on the fly is kinda sus, but Im not sure where we would want to sort it. Maybe we can sort it the first time its been used?